src/plugins/objects/objectmessage.ts

```diff
@@ -706,4 +750,63 @@ export class ObjectMessage {
 
     return size;
   }
+
+  private _decodeMapEntries(
+    mapEntries: Record<string, ObjectsMapEntry<WireObjectData>>,
+    client: BaseClient,
+    format: Utils.Format | undefined,
+  ): Record<string, ObjectsMapEntry<ObjectData>> {
+    return Object.entries(mapEntries).reduce(
+      (acc, v) => {
+        const [key, entry] = v;
+        const decodedData = entry.data ? this._decodeObjectData(entry.data, client, format) : undefined;
+        acc[key] = {
+          ...entry,
+          data: decodedData,
+        };
+        return acc;
+      },
+      {} as Record<string, ObjectsMapEntry<ObjectData>>,
+    );
+  }
+
+  /** @spec OD5 */
+  private _decodeObjectData(
+    objectData: WireObjectData,
+    client: BaseClient,
+    format: Utils.Format | undefined,
+  ): ObjectData {
+    // TODO: support decoding JSON objects stored as a JSON string with an encoding of "json"
+    // https://ably.atlassian.net/browse/PUB-1667
+    // currently we check only the "bytes" field:
+    // - if connection is msgpack - "bytes" was received as msgpack encoded bytes, no need to decode, it's already a buffer
+    // - if connection is json - "bytes" was received as a base64 string, need to decode it to a buffer
+
+    try {
+      let decodedBytes: Bufferlike | undefined;
+      if (format === 'msgpack') {
+        // OD5a1 - connection is using msgpack protocol, bytes are already a buffer
+        decodedBytes = objectData.bytes as Bufferlike;
+      } else {
+        // OD5b2 - connection is using JSON protocol, Base64-decode bytes value
+        decodedBytes =
+          objectData.bytes != null ? client.Platform.BufferUtils.base64Decode(String(objectData.bytes)) : undefined;
+      }
+
+      return {
+        ...objectData,
+        bytes: decodedBytes,
+      };
+    } catch (error) {
+      client.Logger.logAction(
+        client.logger,
+        client.Logger.LOG_ERROR,
+        'WireObjectMessage._decodeObjectData()',
+        this._utils.inspectError(error),
+      );
+      return {
+        ...objectData,
+      } as ObjectData;
```

Comment:
There seem to be quite a few type assertions (`x as y`) in this file. It's not clear to me at a glance whether they're needed or whether they're just a shortcut to allow us to write less code. Do you have any thoughts on this? The less we can bypass compiler checking, the better.